### PR TITLE
bugfix: should force convert weight to number since may cause deadloop when weight is string type "0".

### PR DESCRIPTION
**Pull request by doujiang24**

### CODE REVIEW

File: lib/resty/balancer/utils.lua [hunk 1]
```diff
@@ -11,27 +11,10 @@ do
...
-            end
-
-            return newnodes
-        end
-    end
-end
-_M.copy = copy
 
-local nkeys
+local nkeys, jit_nkeys
```
Comment (agentzh) [inline lib/resty/balancer/utils.lua]: It's confusing to have both `nkeys` and `jit_nkeys`.

**Review by agentzh — CHANGES_REQUESTED** [review]

File: lib/resty/balancer/utils.lua [hunk 2]
```diff
@@ -43,9 +26,22 @@ do
...
+        jit_nkeys = nkeys
     end
 end
-
 _M.nkeys = nkeys
 
+
+function _M.copy(nodes)
+    local newnodes = new_tab(0, jit_nkeys and jit_nkeys(nodes) or 4)
+    for id, weight in pairs(nodes) do
```
Comment (agentzh) [inline lib/resty/balancer/utils.lua:39]: Note that this loop cannot be JIT compiled.

File: lib/resty/balancer/utils.lua [hunk 3]
```diff
@@ -43,9 +26,22 @@ do
...
+    else
+        jit_nkeys = nkeys
     end
 end
-
 _M.nkeys = nkeys
 
+
+function _M.copy(nodes)
+    local newnodes = new_tab(0, jit_nkeys and jit_nkeys(nodes) or 4)
```
Comment (agentzh) [inline lib/resty/balancer/utils.lua]: Can we recycle the new table? Is it on the hot code path?

**Review by agentzh — CHANGES_REQUESTED** [review]

File: lib/resty/balancer/utils.lua (hunk 3 above)
Comment (doujiang24) [reply lib/resty/balancer/utils.lua]: it should not on the hot code path.

File: lib/resty/balancer/utils.lua (hunk 2 above)
Comment (doujiang24) [reply lib/resty/balancer/utils.lua:39]: Does anything need to change? we can not avoid `pairs` here.

File: lib/resty/balancer/utils.lua (hunk 1 above)
Comment (doujiang24) [reply lib/resty/balancer/utils.lua]: how about `nkeys` and `raw_jit_nkeys`?

Comment (agentzh) [reply lib/resty/balancer/utils.lua]: what is the difference here? It's still confusing.

File: lib/resty/balancer/utils.lua (hunk 2 above)
Comment (agentzh) [reply lib/resty/balancer/utils.lua:39]: If it's not on the hot code path, then it's fine.

**Review by zhuizhuhaomeng — APPROVED** [review]

**Pull request merged**